gh-44968: Add "Reload from Disk" feature to IDLE

ashm-dev opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNNN: Summary of the changes made
```

Where: gh-NNNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNNNN)
```

Where: [X.Y] is the branch name, for example: [3.13].

GH-NNNNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-44968 -->
* Issue: gh-44968
<!-- /gh-issue-number -->

ashm-dev committed:
gh-44968: Add "Reload from Disk" feature to IDLE

ashm-dev committed:
Use builtins.open for file operations in tests

sharktide commented on Lib/idlelib/iomenu.py:
```diff
@@ -237,6 +239,49 @@ def save_a_copy(self, event):
+        """
+        if not self.filename:
+            messagebox.showinfo(
+                "No File",
```
```suggestion
                “File Not Found",
```
I think File Not Found would be better because it is a more descriptive version of “No File” in this case

sharktide reviewed (commented):
I also don’t think we should restore the cursor position since a file’s line numbers could also change when it is edited and it could cause confusion.

ashm-dev committed:
Update Lib/idlelib/iomenu.py

Co-authored-by: R.C.M <sharktidedev@gmail.com>

ashm-dev committed:
Simplify file reload by removing cursor restoration

Removed cursor position restoration logic during file reload.

ashm-dev committed:
Merge branch 'main' into idle

ashm-dev committed:
Fix typo in 'File Not Found' message

ashm-dev committed:
Update error message in test for file reload

picnixz commented on Lib/idlelib/idle_test/test_iomenu.py:
```diff
@@ -2,6 +2,7 @@
 
 from idlelib import iomenu
 import unittest
+import builtins
```
Import it before unittest.

picnixz commented on Lib/idlelib/idle_test/test_iomenu.py:
```diff
@@ -57,6 +58,130 @@ def test_fixnewlines_end(self):
+        # Test reload when no file is associated
+        import tempfile
+        import os
+        from unittest.mock import Mock
```
Why using local imports? just make them top-level it does not hurt.

picnixz commented on Lib/idlelib/idle_test/test_iomenu.py:
```diff
@@ -57,6 +58,130 @@ def test_fixnewlines_end(self):
+        showinfo_called = []
+        def mock_showinfo(*args, **kwargs):
+            showinfo_called.append((args, kwargs))
+        iomenu.messagebox.showinfo = mock_showinfo
```
Mock the function and use call_args from Mock objects.

picnixz commented on Lib/idlelib/idle_test/test_iomenu.py:
```diff
@@ -57,6 +58,130 @@ def test_fixnewlines_end(self):
+            self.assertEqual(result, "break")
+            self.assertEqual(text.get('1.0', 'end-1c'), "# Modified content\n")
+        finally:
+            os.unlink(temp_filename)
```
Use self.addCleanup after creating the tempfile instead of try-finally.

picnixz commented on Lib/idlelib/idle_test/test_iomenu.py:
```diff
@@ -57,6 +58,130 @@ def test_fixnewlines_end(self):
+        io = self.io
+        text = io.editwin.text
+
+        # Create a temporary file
```
remove this redundant comment

picnixz commented on Lib/idlelib/idle_test/test_iomenu.py:
```diff
@@ -57,6 +58,130 @@ def test_fixnewlines_end(self):
+
+            # Mock askokcancel to return False (cancel)
+            orig_askokcancel = iomenu.messagebox.askokcancel
+            iomenu.messagebox.askokcancel = lambda *args, **kwargs: False
```
Use return_value + mock.patch()

picnixz reviewed (dismissed):
Please rely more on unittest.mock.patch() as well as mock features directly. In addition add a What's New entry in addition to the NEWS entry.

ashm-dev committed:
fix import

ashm-dev committed:
Merge remote-tracking branch 'origin/idle' into idle

ashm-dev committed:
fix: move local imports to global

ashm-dev committed:
test_iomenu: use unittest.mock.patch and call_args

ashm-dev committed:
IDLE tests: use addCleanup for tempfile removal in test_iomenu

ashm-dev committed:
remove comments

ashm-dev committed:
refactor(tests): Use mock.patch for messagebox dialogs

ashm-dev committed:
add news

ashm-dev committed:
rewrite tests

ashm-dev committed:
Merge branch 'main' into idle

sharktide commented on Doc/whatsnew/3.15.rst:
```diff
@@ -533,6 +533,12 @@ http.cookies
+
+*   Add a "Reload from Disk" item to the File menu. This allows discarding
+    unsaved changes and reloading the current version of the file from the disk.
+    Patch by Shamil Abdulaev. (Contributed by Shamil Abdulaev in :gh:`44968`.)
```
```suggestion
    (Contributed by Shamil Abdulaev in :gh:`44968`.)
```
Unnecessary; your name is still right after it

ashm-dev committed:
Update Doc/whatsnew/3.15.rst

Co-authored-by: R.C.M <sharktidedev@gmail.com>

picnixz commented on Doc/whatsnew/3.15.rst:
```diff
@@ -533,6 +533,12 @@ http.cookies
+
+*   Add a "Reload from Disk" item to the File menu. This allows discarding
+    unsaved changes and reloading the current version of the file from the disk.
+    (Contributed by Shamil Abdulaev in :gh:`44968`.)
```
```suggestion


idlelib
-------

*   Add a "Reload from Disk" item to the File menu. This allows discarding
    unsaved changes and reloading the current version of the file from the disk.
    (Contributed by Shamil Abdulaev in :gh:`44968`.)

```

The module is idlelib but I'm not sure we need to document the module itself. Can you check in previous What's New files (3.12, 3.13 or older) where we documented such changes please?

ashm-dev replied:
Previous versions did not mention idlelib in whatsnew.

picnixz commented on Lib/idlelib/idle_test/test_iomenu.py:
```diff
@@ -57,6 +58,83 @@ def test_fixnewlines_end(self):
+        m.side_effect = [
+            mock_open(read_data=original_content).return_value,
+            mock_open(read_data=modified_content).return_value,
+        ]
```
Do you really need such complicate stuff? can't you just have `m.side_effect = [original_content, modified_content]`?

ashm-dev replied:
The suggested change won't work because `open()` must return a file-like object that has a `.read()` method, not just a raw string. The current code uses `mock_open` to create exactly that fake file object.

picnixz commented on Lib/idlelib/mainmenu.py:
```diff
@@ -31,6 +31,7 @@
    ('_Save', '<<save-window>>'),
    ('Save _As...', '<<save-window-as-file>>'),
    ('Save Cop_y As...', '<<save-copy-of-window-as-file>>'),
+   ('Re_load from Disk', '<<reload-window>>'),
```
I think the `R` for `Reload` is more important than the `L`

picnixz commented on Misc/NEWS.d/next/IDLE/2025-11-14-20-58-55.gh-issue-44968.rL9dK3.rst:
```diff
@@ -0,0 +1,4 @@
+Add "Reload from Disk" menu item to IDLE's File menu. This allows users to
+easily reload a file from disk, discarding any unsaved changes in the editor.
+The feature is particularly useful when working with version control systems
+or when external tools modify files. Patch by Shamil Abdulaev.
```
```suggestion
Patch by Shamil Abdulaev.
```

ashm-dev committed:
Update Misc/NEWS.d/next/IDLE/2025-11-14-20-58-55.gh-issue-44968.rL9dK3.rst

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

ashm-dev committed:
Update Doc/whatsnew/3.15.rst

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

picnixz commented on Doc/whatsnew/3.15.rst:
```diff
@@ -534,6 +534,14 @@ http.cookies
+
+*   Add a "Reload from Disk" item to the File menu. This allows discarding
+    unsaved changes and reloading the current version of the file from the disk.
+    (Contributed by Shamil Abdulaev in :gh:`44968`.)
```
```suggestion
*  Add a "Reload from Disk" item to the File menu. This allows discarding
   unsaved changes and reloading the current version of the file from the disk.
   (Contributed by Shamil Abdulaev in :gh:`44968`.)
```

ashm-dev committed:
Update Doc/whatsnew/3.15.rst

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

ashm-dev committed:
fix

ashm-dev committed:
Merge remote-tracking branch 'origin/idle' into idle

ashm-dev committed:
fix tsan

ashm-dev committed:
fix tests

ashm-dev committed:
Merge branch 'main' into idle

ashm-dev committed:
Merge branch 'main' into idle

hugovk committed:
Python 3.15.0a2

ashm-dev committed:
Merge branch 'main' into idle

sharktide commented on Lib/idlelib/idle_test/test_iomenu.py:
```diff
@@ -32,10 +35,18 @@ def tearDownClass(cls):
         cls.root.update_idletasks()
         for id in cls.root.tk.call('after', 'info'):
-            cls.root.after_cancel(id)  # Need for EditorWindow.
+            cls.root.after_cancel(id)
```
```suggestion
            cls.root.after_cancel(id)  # Need for EditorWindow.
```

Unrelated change

sharktide commented on Lib/idlelib/idle_test/test_iomenu.py:
```diff
@@ -45,17 +56,88 @@ def test_fixnewlines_end(self):
         shelltext = '>>> if 1'
         self.editwin.get_prompt_text = Func(result=shelltext)
-        eq(fix(), shelltext)  # Get... call and '\n' not added.
+        eq(fix(), shelltext)
```
```suggestion
        eq(fix(), shelltext)   # Get... call and '\n' not added.
```

Unrelated change

freakboy3742 committed:
gh-141691: Apply ruff rules to Apple folder. (#141694)

Add ruff configuration to run over the Apple build tooling and testbed script.

ThierryMT committed:
gh-140381: Increase slow_fibonacci call frequency in test_profiling (#140673)

LindaSummer committed:
gh-139103: fix free-threading `dataclass.__init__` perf issue (gh-141596)

The dataclasses `__init__` function is generated dynamically by a call to `exec()` and so doesn't have deferred reference counting enabled. Enable deferred reference counting on functions when assigned as an attribute to type objects to avoid reference count contention when creating dataclass instances.

freakboy3742 committed:
gh-141692: Add a slice-specific lib folder to iOS XCframeworks. (#141693)

Modifies the iOS XCframework to include a lib folder for each slice that
contains a symlinked version of the libPython dynamic library.

ashm-dev committed:
Update Lib/idlelib/idle_test/test_iomenu.py

Co-authored-by: R.C.M <sharktidedev@gmail.com>

ashm-dev committed:
Update Lib/idlelib/idle_test/test_iomenu.py

Co-authored-by: R.C.M <sharktidedev@gmail.com>

ayappanec committed:
gh-141659: Fix bad file descriptor error in subprocess on AIX (GH-141660)

/proc/self does not exist on AIX.

hugovk committed:
Post 3.15.0a2

markshannon committed:
GH-139653: Only raise an exception (or fatal error) when the stack pointer is about to overflow the stack. (GH-141711)

Only raises if the stack pointer is both below the limit *and* above the stack base.
This prevents false positives for user-space threads, as the stack pointer will be outside those bounds
if the stack has been swapped.

guoci committed:
Correct class name from PullDom to PullDOM (#141207)

da-woods committed:
gh-141726: Add PyDict_SetDefaultRef() to the Stable ABI (#141727)

Co-authored-by: Stan Ulbrych <89152624+StanFromIreland@users.noreply.github.com>

serhiy-storchaka committed:
gh-140875: Fix handling of unclosed charrefs before EOF in HTMLParser (GH-140904)

encukou committed:
gh-135953: Avoid BytesWarning when sampling profiler tests fail (GH-141719)

With `-bb`, this failed on BytesWarning: str() on a bytes instance

hroncok committed:
gh-141570: can_colorize: Expect fileno() to raise OSError, as documented (#141716)

In Fedora, we've been given a slightly incomplete reproducer for a problematic
Python 3.14 color-related change in argparse that leads to an exception when
Python is used from mod_wsgi: https://bugzilla.redhat.com/2414940

mod_wsgi replaces sys.stdout with a custom object that raises OSError on .fileno():

https://github.com/GrahamDumpleton/mod_wsgi/blob/8460dbfcd5c7108892b3cde9fab7cbc1caa27886/src/server/wsgi_logger.c#L434-L440

This should be supported, as the documentation of fileno explicitly says:

> An OSError is raised if the IO object does not use a file descriptor.

https://docs.python.org/3.14/library/io.html#io.IOBase.fileno

The previously expected exception inherits from OSError,
so it is still expected.

Fixes https://github.com/python/cpython/issues/141570

Co-authored-by: Cody Maloney <cmaloney@users.noreply.github.com>
Co-authored-by: Victor Stinner <vstinner@python.org>

anordin95 committed:
gh-138072: Small clarifications and phrasing improvements to asyncio HOWTO (#138073)

* - Small clarifications and phrasing improvements

* nit

* Apply suggestions from code review

Co-authored-by: Peter Bierma <zintensitydev@gmail.com>

* clarify event loops when multi threading.

* nit

* Update Doc/howto/a-conceptual-overview-of-asyncio.rst

Co-authored-by: Peter Bierma <zintensitydev@gmail.com>

* nit

* nit

* phrasing for threads & event loops.

* revert changes to event-loop/thread discussion.

* sentence case consistencty.

* slight re-arrange.

* Sentence case consistency.

* tweak language. non-negligible

---------

Co-authored-by: Peter Bierma <zintensitydev@gmail.com>

brandtbucher committed:
GH-140638: Add a GC "duration" stat (GH-141720)

serhiy-storchaka committed:
gh-41779: Allow defining the __dict__ and __weakref__ __slots__ for any class (GH-141755)

alexcrichton committed:
gh-141172: Update to wasi-sdk-29 (GH-141684)

This commit updates CI and configuration from wasi-sdk-25 to wasi-sdk-29
which was released recently. This notably includes stubs for pthreads
which all return errors, so some adjustment in logic is necessary to
retain knowledge that WASI cannot yet spawn threads for example.

This additionally increases the wasm stack allowance to 32MiB from 16MiB
to accomodate the `test_recursive_pickle` test in the
`test_functools.py` file. It looks like the Clang/LLVM update that
happened in wasi-sdk-29 relative to wasi-sdk-25 is likely the cause of
this where presumably functions have more locals than before and/or a
slightly adjusted stack space requirement which overflows the stack.

ramalho committed:
gh-141721: Improve docstring for LastUpdatedOrderedDict example (gh141724)

StanFromIreland committed:
gh-141004: Document `Py_LOCAL` and `Py_LOCAL_INLINE` (GH-141725)

ayappanec committed:
Move the NEWS entry for gh-141659 to the correct section (GH-141744)

gvanrossum committed:
Remove GvR from '_stdauthor' in Doc/conf.py (#141765)

Co-authored-by: Guido van Rossum <gvanrossum@microsoft.com>

brettcannon committed:
Simplify setting CLI options for WASI builds (GH-141769)

This introduces a Wasmtime configuration file to get some CLI options out of the code for easier manipulation. It also allows for easier tweaking after the Makefile is generated.

As well, cut back on the flexibility of specifying HOSTRUNNER for simpler code. The flexibility was never used and so it didn't make sense to keep it around.

guoci committed:
[Docs] Fix typo in bdb: `is_skipped_line` to `is_skipped_module` (#141771)

gaogaotiantian committed:
gh-141615: Check stdin instead of stdout for use_rawinput in pdb (#141616)

dr-carlos committed:
gh-141489: Simplify closure/freevar iteration in `annotationlib._build_closure()` (#141490)

cmaloney committed:
gh-139871: Optimize small takes in bytearray.take_bytes (GH-141741)

When less than half the buffer is taken just copy that small part out
rather than doing a big alloc + memmove + big shrink.

Bobronium committed:
gh-141757: Comply with pickle protocol in test_reduce_5tuple (GH-141759)

The 5th element of the __reduce__ tuple should be an iterator.

ZeroIntensity committed:
gh-141004: Document `Py_MakePendingCalls` (GH-141137)

Co-authored-by: Victor Stinner <vstinner@python.org>

ZeroIntensity committed:
gh-141004: Document missing generator APIs (GH-141409)

Co-authored-by: Victor Stinner <vstinner@python.org>

ZeroIntensity committed:
gh-141004: Document missing `PyDateTime*` APIs (GH-141543)

Co-authored-by: Stan Ulbrych <89152624+StanFromIreland@users.noreply.github.com>
Co-authored-by: Victor Stinner <vstinner@python.org>

prithviraj-chaudhuri committed:
gh-140042: Removing unsafe call to sqlite3_shutdown (GH-141690)

vstinner committed:
gh-141784: Fix _remote_debugging_module.c compilation on 32-bit Linux (#141796)

Include Python.h before system headers to make sure that
_remote_debugging_module.c uses the same types (ABI) than Python.

cmaloney committed:
gh-60107: Remove a copy from RawIOBase.read (#141532)

If the underlying I/O class keeps a reference to the memory, raise BufferError.

Co-authored-by: Victor Stinner <vstinner@python.org>

pablogsal committed:
gh-141645: Add a TUI mode to the new tachyon profiler (#141646)

serhiy-storchaka committed:
gh-138697: Fix inferring dest from a single-dash long option in argparse (#138699)

* gh-138697: Fix inferring dest from a single-dash long option in argparse

If a short option and a single-dash long option are passed to add_argument(),
dest is now inferred from the single-dash long option.

* Make double-dash options taking priority over single-dash long options.

---------

Co-authored-by: Savannah Ostrowski <savannah@python.org>

pablogsal committed:
gh-141645: Refactor tachyon's live TUI tests to not use private fields (#141806)

michaelkiper committed:
gh-141729: Doc PyTypeObject.tp_vectorcall grammar fix (#141730)

[3.14] Doc PyTypeObject.tp_vectorcall grammar fix

Co-authored-by: makiper <michael.kiper@jpl.nasa.gov>

sergey-miryanov committed:
GH-139727: Add --enable-stackref-debug flag to build.bat to enable Py_STACKREF_DEBUG build on Windows (GH-139728)

pablogsal committed:
gh-141645: Add profiling/sampling/live_collector to the makefile and remove spurious file (#141820)

kumaraditya303 committed:
gh-140795: fetch thread state once on fast path for critical sections (#141406)

pareshjoshij committed:
gh-141612: improve `test_trampoline_works_with_forks` coverage (#141613)

yoney committed:
gh-116738: Make csv module thread-safe (gh-141365)

Added a critical section to protect the states of `ReaderObj` and `WriterObj` in the free-threading build. Without the critical sections, both new free-threading tests were crashing.

efimov-mikhail committed:
gh-141498: Change backoff counter to use prime numbers instead of powers of 2 (GH-141591)

cmaloney committed:
gh-129559: add `bytearray.resize` thread safety test for free-threading  (#141739)

colesbury committed:
gh-137422: Fix race condition in PyImport_AddModuleRef (gh-141822)

pablogsal committed:
gh-140381: Make test_profiling tests deterministic to fix flakiness (#141731)

picnixz commented on Lib/idlelib/idle_test/test_iomenu.py:
```diff
@@ -45,17 +56,88 @@ def test_fixnewlines_end(self):
         shelltext = '>>> if 1'
         self.editwin.get_prompt_text = Func(result=shelltext)
-        eq(fix(), shelltext)  # Get... call and '\n' not added.
+        eq(fix(), shelltext)   # Get... call and '\n' not added.
```
Ok, this will be the last time I'm going to say this but please **never change unrelated code**. Do not use an autoformatter, or configure your IDE so that it doesn't autoformat on save, or ask your LLM not to do any change.

sharktide replied:
```suggestion
        eq(fix(), shelltext)  # Get... call and '\n' not added.
```

picnixz commented on Doc/whatsnew/3.15.rst:
```diff
@@ -534,6 +534,14 @@ http.cookies
   (Contributed by Nick Burns and Senthil Kumaran in :gh:`92936`.)
 
 
+idlelib
```
@terryjreedy For IDLE changes, do we add it here or somewhere else in general?

picnixz commented on Lib/idlelib/idle_test/test_iomenu.py:
```diff
@@ -45,17 +56,88 @@ def test_fixnewlines_end(self):
+        io = self.io
+        io.filename = None
+
+        with patch('idlelib.iomenu.messagebox.showinfo') as mock_showinfo:
+            result = io.reload(None)
+            self.assertEqual(result, "break")
+            mock_showinfo.assert_called_once()
+            args, kwargs = mock_showinfo.call_args
+            self.assertIn("File Not Found", args[0])
```
```suggestion
        with patch('idlelib.iomenu.messagebox.showinfo') as mock_showinfo:
            result = io.reload(None)
            self.assertEqual(result, "break")
        mock_showinfo.assert_called_once()
        args, _ = mock_showinfo.call_args
        self.assertIn("File Not Found", args[0])
```

picnixz commented on Lib/idlelib/idle_test/test_iomenu.py:
```diff
@@ -45,17 +56,88 @@ def test_fixnewlines_end(self):
+            io.loadfile(io.filename)
+            self.assertEqual(text.get('1.0', 'end-1c'), original_content)
+
+            with builtins.open(filename, 'w', encoding='utf-8') as f:
```
Why using builtins.open now that you have `tokenize_open`?

sharktide replied:
I'm pretty sure tokenize_open can only open the file in read mode here

picnixz replied:
It's not about tokenize_open. It's about just using the plain `open`. There is no need to use `builtins.open` now that `open` is not shadowed.

picnixz commented on Lib/idlelib/idle_test/test_iomenu.py:
```diff
@@ -36,6 +39,14 @@ def tearDownClass(cls):
         del cls.root
 
+    def _create_tempfile(self, content: str) -> str:
+        fd, filename = tempfile.mkstemp(suffix='.py')
+        os.close(fd)
+        self.addCleanup(os.unlink, filename)
+        with builtins.open(filename, 'w', encoding='utf-8') as f:
+            f.write(content)
+        return filename
```
```suggestion
        filename = tempfile.mktemp()
        self.addCleanup(os_helper.unlink, filename)
        with open(filename, "w") as f:
            f.write(content)
        return filename
```

We don't care about secure temporary file for a test.

picnixz commented on Lib/idlelib/idle_test/test_iomenu.py:
```diff
@@ -36,6 +39,14 @@ def tearDownClass(cls):
         cls.root.destroy()
         del cls.root
 
+    def _create_tempfile(self, content: str) -> str:
```
Don't use type annotations.

sharktide replied:
```suggestion
    def _create_tempfile(self, content):
```

picnixz commented on Lib/idlelib/idle_test/test_iomenu.py:
```diff
@@ -45,17 +56,88 @@ def test_fixnewlines_end(self):
+        text.insert('end', "\n# Unsaved change")
+        io.set_saved(False)
+
+        with patch('idlelib.iomenu.messagebox.askokcancel', return_value=False) as mock_ask:
```
Please wrap lines under 80 chars.

sharktide replied:
```suggestion
        with patch('idlelib.iomenu.messagebox.askokcancel',
                return_value=False) as mock_ask:
```

picnixz reviewed (changes requested):
Please, your PRs always take me long to review because there are unrelated changes, or un-necessary things that are introduced. It's really not the first time. Carefully read the devguide and honestly, avoid using an LLM. If you are worried about NEWS writing, leave it blank and ask someone else to write it for you. But for the code, there are many places where I feel it's been LLM-generated. In general:

- Never remove code that is not related to the issue you are fixing.
- Never add type annotations.
- Never make unrelated changes.
- Read the devguide.
- Don't add trivial comments. Usually this is a sign that an LLM has been used.
- Follow PEP-8. In particular, wrap your code under the 80-char mark.

sharktide commented on Misc/NEWS.d/next/IDLE/2025-11-14-20-58-55.gh-issue-44968.rL9dK3.rst:
```diff
@@ -0,0 +1,3 @@
+Add "Reload from Disk" menu item to IDLE's File menu. This allows users to
+easily reload a file from disk, discarding any unsaved changes in the editor.
```
```suggestion
reload a file from disk, discarding unsaved changes in the editor.
```
Remove unnecessary words

sharktide commented on Lib/idlelib/iomenu.py:
```diff
@@ -237,6 +239,35 @@ def save_a_copy(self, event):
+                self.text.focus_set()
+                return "break"
+
+        # Reload the file
```
```suggestion
```
Trivial comment

sharktide reviewed (commented):
Added commit suggestions based on picnixz's comments, also added one suggestion for the news entry

Yzi-Li committed:
gh-141833: Remove the bad lines in `test_attr_promotion_failure` (GH-141841)

Rosuav committed:
gh-141817: Add IPV6_HDRINCL constant to the socket module (#141818)

Co-authored-by: Victor Stinner <vstinner@python.org>
Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

donbarbos committed:
gh-131178: Add tests for `pickletools` command-line interface (#131287)

davidben committed:
gh-141801: Use accessors for ASN1_STRING fields in libssl (GH-141802)

* gh-141801: Use accessors for ASN1_STRING fields

While ASN1_STRING is currently exposed, it is better to use the
accessors. See https://github.com/openssl/openssl/issues/29117 where, if
the type were opaque, OpenSSL's X509 objects could be much more
memory-efficient.

* Update Modules/_ssl.c

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

* Update Modules/_ssl.c

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

---------

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

ahesford committed:
gh-141600: Fix musl version detection on Void Linux (GH-141602)

gpshead committed:
gh-98552: flush std streams in the multiprocessing forkserver before fork (#141849)

* flush std streams in the multiprocessing forkserver before fork

* NEWS

LamentXU123 committed:
gh-101100: Fix sphinx warnings in `library/unittest.rst` (#140109)

Co-authored-by: Hugo van Kemenade <1324225+hugovk@users.noreply.github.com>

AA-Turner committed:
GH-101100: Resolve reference warnings in library/stdtypes.rst (#138420)

serhiy-storchaka committed:
gh-138525: Support single-dash long options and prefix_chars in BooleanOptionalAction (GH-138692)

-nofoo is generated for -foo.
++no-foo is generated for ++foo.
/nofoo is generated for /foo.

brandtbucher committed:
GH-140638: Add a GC "candidates" stat (GH-141814)

hugovk committed:
gh-141679: Add colour to defaults in `argparse` help (#141680)

Co-authored-by: Savannah Ostrowski <savannah@python.org>

gpshead committed:
[security] Add a cooldown period to dependabot (GH-141866)

See https://blog.yossarian.net/2025/11/21/We-should-all-be-using-dependency-cooldowns and the description in the comment.

gpshead committed:
gh-98552: Revert (unneeded, already done elsewhere) "flush std streams in the multiprocessing forkserver before fork (#141849)" (#141871)

Revert (unneeded, already done elsewhere) "gh-98552: flush std streams in the multiprocessing forkserver before fork (#141849)"

This reverts commit 58badb1711e12b6e8b5240ab96cdd01b95012de7.

colesbury committed:
gh-120158: Fix inconsistent monitoring state when setting events too frequently (gh-141845)

If we overflowed the global version counter (i.e., after 2*24 calls to
`_PyMonitoring_SetEvents`), we bailed out after setting global monitoring
events but before instrumenting code objects, which led to assertion errors
later on.

Also add a `time.sleep()` to `test_free_threading.test_monitoring` to avoid
overflowing the global version counter.

yihong0618 committed:
Drop three unused imports (#141875)

StanFromIreland committed:
gh-141865: Fix translation of strings in `changes.py` extension (#141882)

srepmub committed:
math.dist documentation: add 'strict=True' to equivalent code (#141819)

duaneg committed:
gh-135335: Simplify preload regression test using __main__ (GH-138686)

Simplify preload regression test using `__main__`

With the fix for gh-126631 `__main__` modules can be preloaded and the regression
test for gh-135335 can be simplified to just use a self-contained script rather
than requiring a module.

Note this assumes and implicitly tests that `__main__` is preloaded by default.

pablogsal committed:
gh-138122: Refactor the CLI of profiling.sampling into subcommands (#141813)

adorilson committed:
gh-106318: Add example for str.isdecimal() (#137559)

Co-authored-by: Hugo van Kemenade <1324225+hugovk@users.noreply.github.com>

encukou committed:
gh-141780: Make PyModule_FromSlotsAndSpec enable GIL if needed  (GH-141785)

Yashp002 committed:
gh-141004: Document `Py_ARRAY_LENGTH` (GH-141601)

Co-authored-by: Stan Ulbrych <89152624+StanFromIreland@users.noreply.github.com>
Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>
Co-authored-by: Peter Bierma <zintensitydev@gmail.com>

adorilson committed:
gh-106318: Add example for str.isascii() (#137558)

Co-authored-by: Stan Ulbrych <89152624+StanFromIreland@users.noreply.github.com>
Co-authored-by: Hugo van Kemenade <1324225+hugovk@users.noreply.github.com>

anordin95 committed:
gh-138199: Address potential confusion regarding threading in the `asyncio` HOWTO (GH-138200)

adorilson committed:
gh-106318: Add example for str.isalpha() (#137557)

Co-authored-by: Éric <merwok@netwok.org>
Co-authored-by: Hugo van Kemenade <1324225+hugovk@users.noreply.github.com>
Co-authored-by: Victor Stinner <vstinner@python.org>

pablogsal committed:
gh-138122: Fix permission error detection in tachyon's test (#141897)

SkvProgrammer committed:
gh-141831: Document behavior of functools.lru_cache with concurrent access (gh-141869)

adorilson committed:
gh-106318: Add example for str.format() (#137018)

adorilson committed:
gh-106318: Add examples for str.join() (#140315)

Co-authored-by: Stan Ulbrych <89152624+StanFromIreland@users.noreply.github.com>

cmaloney committed:
gh-141863: use `bytearray.take_bytes` in asyncio streams for better performance (#141864)

colesbury committed:
gh-116749: Fix code comment for _PyConfig_GIL_DEFAULT (gh-141804)

colesbury committed:
gh-129441: Fix some flakiness in test_instrumentation (gh-141881)

Most of the `self.assertTrue(self.called)` checks are flaky because
the worker threads may sometimes finish before the main thread calls
`self.during_threads()`.

ZeroIntensity committed:
gh-141004: Document `PyImport_Inittab` (GH-141844)

Co-authored-by: Stan Ulbrych <stan@ulbrych.org>
Co-authored-by: Victor Stinner <vstinner@python.org>

ZeroIntensity committed:
gh-141004: Document stack effect C APIs (GH-141843)

Co-authored-by: Victor Stinner <vstinner@python.org>

Ansuel committed:
gh-141907: Better handle support for SHA3 for test_hashlib (GH-141908)

* test_hashlib: better handle support for SHA3

It's possible that the SSL library supports only SHA3 algo and doesn't
have SHAKE one.

The current test wrongly detect this and set both HASH and HASHXOF to
None expecting to have the extra SHA3 attributes present but this should
only be true for SHAKE algo.

To better handle this, move the HASH condition to a dedicated try-expect
condition and check if HASHXOF is None in the relevant code effectively
checking if SHA3 is supported by the SSL library but SHAKE algo needs to
use the sha3module one.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

* rework the conditional import for all its attrs

---------

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>
Co-authored-by: Gregory P. Smith <greg@krypto.org>

SubbaraoGarlapati committed:
Fix implicit import in `test_monitoring.py` (gh-141795)

sergey-miryanov committed:
GH-141861: Fix invalid memory read in the ENTER_EXECUTOR (GH-141921)

pareshjoshij committed:
gh-141781: Fix pdb.line_prefix binding (#141779)

pablogsal committed:
gh-138122: Split Modules/_remote_debugging_module.c into multiple files (#141934)

gh-1381228: Split Modules/_remote_debugging_module.c into multiple files

XChaitanyaX committed:
gh-140911: Ensure that UserString.index() and UserString.rindex() accept UserString as argument (GH-140945)

encukou committed:
gh-139165: Make Py_SIZE, Py_IS_TYPE,Py_ SET_SIZE regular functions in stable ABI (GH-139166)

* Make Py_{SIZE,IS_TYPE,SET_SIZE} regular functions in stable ABI

Group them together with Py_TYPE & Py_SET_TYPE to cut down
on repetitive preprocessor macros.
Format repetitive definitions in object.c more concisely.

Py_SET_TYPE is still left out of the Limited API.

StanFromIreland committed:
gh-141004: Document `PyOS_mystr(n)icmp` (#141760)

Co-authored-by: Petr Viktorin <encukou@gmail.com>

encukou committed:
gh-141909: Add `PyModuleDef_Slot` and earlier Py_mod_* constants to stable ABI manifest (#141910)

These were added to the limited API in 3.5.
Not including them in `Misc/stable_abi.toml` was a bug.

ZeroIntensity committed:
gh-141004: Document missing `PyThread*` APIs (GH-141810)

Co-authored-by: Victor Stinner <vstinner@python.org>

StanFromIreland committed:
Remove references to ``Misc/ACKS`` from ``CONTRIBUTING.md`` (#141952)

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

StanFromIreland committed:
Remove ``Misc/ACKS`` check from patchcheck, documentation (#141960)

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

guoci committed:
Correct indentation in `stdtypes.rst` (#141957)

encukou committed:
gh-140550: PEP 793 reference documentation (GH-141197)

* gh-140550: PEP 793 reference documentation

Since the PEP calls for soft-deprecation of the existing initialization
function, this reorganizes the relevant docs to put the new way of
doing things first, and de-emphasize the old.

Some bits, like the tutorial, are left out of this patch. (See the
issue for a list.)

itamaro committed:
gh-140011: Delete importdl assertion that prevents importing embedded modules from packages (GH-141605)

sergey-miryanov committed:
GH-141861: Fix TRACE_RECORD if full (GH-141959)

encukou committed:
gh-141909: Correct version where Py_mod_gil was added (GH-141979)
